core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,239 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.util.{Timer, TimerTask}
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rpc.{RpcCallContext, RpcEnv, ThreadSafeRpcEndpoint}
+import org.apache.spark.scheduler.{LiveListenerBus, SparkListener, SparkListenerStageCompleted}
+
+/**
+ * A coordinator that handles all global sync requests from BarrierTaskContext. Each global sync
+ * request is generated by `BarrierTaskContext.barrier()`, and identified by
+ * stageId + stageAttemptId + barrierEpoch. Reply all the blocking global sync requests upon
+ * received all the requests for a group of `barrier()` calls. If the coordinator doesn't collect
+ * enough global sync requests within a configured time, fail all the requests due to timeout.
+ */
+private[spark] class BarrierCoordinator(
+    timeout: Int,
+    listenerBus: LiveListenerBus,
+    override val rpcEnv: RpcEnv) extends ThreadSafeRpcEndpoint with Logging {
+
+  private val timer = new Timer("BarrierCoordinator barrier epoch increment timer")
+
+  private val listener = new SparkListener {
+    override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit = {
+      val stageInfo = stageCompleted.stageInfo
+      // Remove internal data from a finished stage attempt.
+      cleanupSyncRequests(stageInfo.stageId, stageInfo.attemptNumber)
+      barrierEpochByStageIdAndAttempt.remove((stageInfo.stageId, stageInfo.attemptNumber))
+      cancelTimerTask(stageInfo.stageId, stageInfo.attemptNumber)
+    }
+  }
+
+  // Epoch counter for each barrier (stage, attempt).
+  private val barrierEpochByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), Int]
+
+  // Remember all the blocking global sync requests for each barrier (stage, attempt).
+  private val syncRequestsByStageIdAndAttempt =
+    new ConcurrentHashMap[(Int, Int), ArrayBuffer[RpcCallContext]]
+
+  // Remember all the TimerTasks for each barrier (stage, attempt).
+  private val timerTaskByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), TimerTask]
+
+  // Number of tasks for each stage.
+  private val numTasksByStage = new ConcurrentHashMap[Int, Int]
+
+  override def onStart(): Unit = {
+    super.onStart()
+    listenerBus.addToStatusQueue(listener)
+  }
+
+  /**
+   * Get the array of [[RpcCallContext]]s that correspond to a barrier sync request from a stage
+   * attempt.
+   */
+  private def getOrInitSyncRequests(
+      stageId: Int,
+      stageAttemptId: Int,
+      numTasks: Int): ArrayBuffer[RpcCallContext] = {
+    val requests = syncRequestsByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
```

Comment:
I think that this code always allocate `ArrayBuffer()` regardless of existence of a value.
If allocating an array is acceptable, the following code looks simpler.
```
val value = new ArrayBuffer[RpcCallContext](numTasks);
val requests = syncRequestsByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId), value)
if (requests != null) {
  requests
} else {
  value
}
```

If allocating an array is not acceptable, we may need to use `computeIfAbsent` like this
```
private val allocateArrayBuffer = new JFunction[Int, ArrayBuffer[RpcCallContext]] {
  override def apply(numTasks: Int): ArrayBuffer[RpcCallContext]= new ArrayBuffer[RpcCallContext](numTasks)
}

private def getOrInitSyncRequests(...): ArrayBuffer[RpcCallContext] = {
  syncRequestsByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId), allocateArrayBuffer)
}
```